devs/docs/logging.rst

```diff
@@ -0,0 +1,56 @@
+=======
+Logging
+=======
+
+Debug and Trace Logging
+.......................
+
+When working with asynchronous programming, it is often hard to follow the
+execution flow or identify possible bugs. This is especially true for
+distributed systems.
+
+Log messages can make this easier. Therefore, when implementing new features,
+refactoring code, or even just debugging, you are encouraged to add
+(or improve) log messages. This is likely to help you with your current work,
+as well as helping other people with future work.
+
+Such messages should be logged using the ``DEBUG`` or ``TRACE`` log level,
+depending on how low-level the information is. To avoid introducing additional
+performance overheads, when the logging statements include some relatively
+expensive computation or they use more than 10 arguments, they must also be
+conditional on the current level of the logger, e.g.::
+
+  if (logger.isTraceEnabled()) {
+      // Relatively expensive calculations
+      ...
+      logger.trace("My useful log message");
+  }
+
+``isXXXEnabled`` are also relatively expensive so if possible save it to a
```

Comment:
These sorts of function calls (\`\`isFooEnabled\`\`) are relatively expensive themselves, so if possible, save the result to a class attribute during object initialisation.